### DISCUSSION TITLE
How do I download a private repository?

### DISCUSSION THREAD

**Discussion by cyberbellum96** [louislam/dockge#339] [Ask for Help] [answered]
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/dockge/security/policy)

### 📝 Describe your problem

When I try to upload a private repository, I get an error, and this is quite logical.

However, I'm wondering how I can access the private repository? I can't open access to it because it contains confidential information.

### 📝 Error Message(s) or Log

Error response from daemon: pull access denied for cyberbellum96/this-my-repo repository does not exist or may require 'docker login': denied: requested access to the resource is denied

### 🐻 Dockge Version

1.4.1

### 💻 Operating System and Arch

Ubuntu 22.04.3 LTS

### 🌐 Browser

N/A

### 🐋 Docker Version

Docker 24.0.7

**Comment by qroac** [comment, marked as answer]
This worked for me on a docker host:
(example uses gitlab, should adapt to other registries, too)

1. Obtain deploy user credentials.  On a gitlab project, navigate to settings -> repository -> deploy token and create a token with permission to read registry
2. SSH to your host, authenticate as root
3. use `docker login registry.gitlab.com -u USER -p PASSWORD` to authenticate with the created credentials
4. This is important, if you work with different repositories using different deployment credentials:
    1. `open /root/.docker/config.json` for editing
    2. Find the entry for `registry.gitlab.com` and append the path of your image, should look like this: `registry.gitlab.com/cyberbellum96/this-my-repo`
    3. Now this credentials are only used for images under this specific path. That allows you to repeat this procedure for other images using different credentials.
5. Mount the docker config folder to dockge as readonly volume: `$HOME/.docker:/root/.docker:ro`
6. Restart dockge. Now dockge can use the credentials to pull your images
